framework/src/meshgenerators/CutMeshByPlaneGenerator.C

```diff
@@ -0,0 +1,502 @@
+  if (MooseUtils::absoluteFuzzyLessThan(dist, 0.0))
+    return PointPlaneRelationIndex::opposite_plane_normal_side;
+  else if (MooseUtils::absoluteFuzzyGreaterThan(dist, 0.0))
+    return PointPlaneRelationIndex::plane_normal_side;
+  else
+    return PointPlaneRelationIndex::on_plane;
+}
+
+Point
+CutMeshByPlaneGenerator::pointPairPlaneInterception(const Point & point1,
+                                                    const Point & point2,
+                                                    const Point & plane_point,
+                                                    const Point & plane_normal) const
+{
+  const Real dist1 = plane_normal * (point1 - plane_point);
+  const Real dist2 = plane_normal * (point2 - plane_point);
+
+  if (MooseUtils::absoluteFuzzyEqual(dist1, 0.0) || MooseUtils::absoluteFuzzyEqual(dist2, 0.0))
+    mooseError("At least one of the two points are on the plane.");
+  if (MooseUtils::absoluteFuzzyGreaterThan(dist1 * dist2, 0.0))
+    mooseError("The two points are on the same side of the plane.");
+
+  const Real dist_ratio = dist1 / (dist1 - dist2);
+  return point1 + dist_ratio * (point2 - point1);
+}
+
+const Node *
+CutMeshByPlaneGenerator::nonDuplicateNodeCreator(ReplicatedMesh & mesh,
+                                                 std::vector<const Node *> & new_on_plane_nodes,
+                                                 const Point & new_point)
+{
+  for (const auto & new_on_plane_node : new_on_plane_nodes)
+  {
+    if (MooseUtils::absoluteFuzzyEqual((*new_on_plane_node - new_point).norm(), 0.0))
+      return new_on_plane_node;
+  }
+  new_on_plane_nodes.push_back(mesh.add_point(new_point));
+  return new_on_plane_nodes.back();
+}
+
+void
+CutMeshByPlaneGenerator::tet4ElemCutter(
+    ReplicatedMesh & mesh,
+    const std::vector<libMesh::BoundaryInfo::BCTuple> & bdry_side_list,
+    const dof_id_type elem_id,
+    const Point & plane_point,
+    const Point & plane_normal,
+    const subdomain_id_type & block_id_to_remove,
+    std::vector<const Node *> & new_on_plane_nodes)
+{
+  // Build boundary information of the mesh
+  BoundaryInfo & boundary_info = mesh.get_boundary_info();
+  // Create a list of sidesets involving the element to be split
+  // It might be complex to assign the boundary id to the new elements
+  // In TET4, none of the four faces have the same normal vector
+  // So we are using the normal vector to identify the faces that
+  // need to be assigned the same boundary id
+  std::vector<Point> elem_side_normal_list;
+  elem_side_normal_list.resize(4);
+  for (const auto i : make_range(4))
+  {
+    auto elem_side = mesh.elem_ptr(elem_id)->side_ptr(i);
+    elem_side_normal_list[i] = (*elem_side->node_ptr(1) - *elem_side->node_ptr(0))
+                                   .cross(*elem_side->node_ptr(2) - *elem_side->node_ptr(1))
+                                   .unit();
+  }
+
+  std::vector<std::vector<boundary_id_type>> elem_side_list;
```

Comment:
add a comment

this is a little expensive to loop over every bc_tuple on the mesh to find the elements they are part of 
I think there is an option to sort the side list. then we could access it faster ?

Reply:
I will create a method for it as it is used for four times